### PR TITLE
Add initial draft of @wrap.check decorator

### PR DESCRIPTION
**Pull request by kaidokert**
Proposed way of doing  #282

### CODE REVIEW

File: docs/wrapping.rst [hunk 1]
```diff
@@ -147,5 +147,24 @@ To avoid the conversion of an argument or return value, use None
...
+
+.. doctest::
+
+    >>> mypp = ureg.check('[length]')(pendulum_period)
+
+In the decorator format:
+
+.. doctest::
+
+    >>>@ureg.wraps('[length]')
```
Comment (hgrecco) [inline docs/wrapping.rst]: this should be `check`

**Comment by kaidokert** [comment]
Whoops i failed python3 really bad here. Will go back and fix

File: pint/unit.py [hunk 2]
```diff
@@ -1145,6 +1145,36 @@ def wrapper(*values, **kw):
...
+        """
+        dimensions = [self.get_dimensionality(dim) for dim in args]
+
+        def decorator(func):
+            assigned = tuple(attr for attr in functools.WRAPPER_ASSIGNMENTS if hasattr(func, attr))
+            updated = tuple(attr for attr in functools.WRAPPER_UPDATES if hasattr(func, attr))
+
+            @functools.wraps(func, assigned=assigned, updated=updated)
+            def wrapper(*values, **kwargs):
+                for dim, value in itertools.izip_longest(dimensions, values):
```
Comment (hgrecco) [inline pint/unit.py]: Two things about this.
1) izip_longest does not exist in Python 3+ (it is called zip_longest). Please add it via the compat module
2) In `wraps` we used zip, but actually I think that zip_longest is better. Would you mind submitting another PR fixing that.

File: pint/unit.py [hunk 3]
```diff
@@ -1145,6 +1145,36 @@ def wrapper(*values, **kw):
...
+                    if dim and value.dimensionality != dim:
+                        raise TypeError(
```
Comment (hgrecco) [inline pint/unit.py]: I think that a `DimensionlityError` would be a better choice

Comment (kaidokert) [inline pint/unit.py]: Yeah i thought about this too, wasnt sure.

**Comment by kaidokert** [comment]
Addressed most of the comments. I'll PR use of zip_longest as well after this. 
DimensionalityError that is thrown looks kinda funny now..
DimensionalityError: Cannot convert from '12 parsec' ([length]) to 'a quantity of' ([length] / [time])
But i think that would be mostly OK and understandable.

**Comment by hgrecco** [comment]
This is merged but somehow GitHub is not recognizing it.

**Pull request closed without merging**